File: javascript/treap/Naive.js
#!/usr/bin/env node

class Node {
    constructor(x) {
        this.x = x
        this.y = Math.random()
        this.left = null
        this.right = null
    }
}


function merge(lower, greater) {
    if (lower === null) {
        return greater
    }

    if (greater === null) {
        return lower
    }

    if (lower.y < greater.y) {
        lower.right = merge(lower.right, greater)

        return lower
    } else {
        greater.left = merge(lower, greater.left)

        return greater
    }
}

function splitBinary(orig, value) {
    if (orig === null) {
        return [null, null]
    }

    if (orig.x < value) {
        const splitPair = splitBinary(orig.right, value)
        orig.right = splitPair[0]

        return [orig, splitPair[1]]
    } else {
        const splitPair = splitBinary(orig.left, value)
        orig.left = splitPair[1]

        return [splitPair[0], orig]
    }
}

function merge3(lower, equal, greater) {
    return merge(merge(lower, equal), greater)
}


class SplitResult {
    constructor(lower, equal, greater) {
        this.lower = lower
        this.equal = equal
        this.greater = greater
    }
}


function split(orig, value) {
    const [lower, equalGreater] = splitBinary(orig, value)
    const [equal, greater] = splitBinary(equalGreater, value + 1)

    return new SplitResult(lower, equal, greater)
}


class Tree {
    constructor() {
        this.root = null
    }

    hasValue(x) {
        const splited = split(this.root, x)
        const res = splited.equal !== null
        this.root = merge3(splited.lower, splited.equal, splited.greater)

        return res
    }

    insert(x) {
        const splited = split(this.root, x)

        if (splited.equal === null) {
            splited.equal = new Node(x)
        }

        this.root = merge3(splited.lower, splited.equal, splited.greater)
    }

    erase(x) {
        const splited = split(this.root, x)
        this.root = merge(splited.lower, splited.greater)
    }
}

function main() {
    const tree = new Tree()
    let cur = 5
    let res = 0

    for (let i = 1; i < 1000000; ++i) {
        let a = i % 3
        cur = (cur * 57 + 43) % 10007
        if (a === 0) {
            tree.insert(cur)
        } else if (a === 1) {
            tree.erase(cur)
        } else if (a == 2) {
            res += tree.hasValue(cur) ? 1 : 0
        }
    }
    console.log(res)
}

(async function() {
    const startTimeMs = new Date().getTime();
      
    main();

    const endTimeMs = new Date().getTime();
    const durationMs = endTimeMs - startTimeMs;
    console.log(`Execution time: ${durationMs}ms`);
})();

File: ruby/treap/Naive.rb
#!/bin/env ruby

class Node
  attr_accessor :x, :y, :left, :right

  def initialize(x)
    @x = x
    @y = rand()
    @left = nil
    @right = nil
  end
end

def merge(lower, greater)
  if lower.nil?
    return greater
  elsif greater.nil?
    return lower
  elsif lower.y < greater.y
    lower.right = merge(lower.right, greater)
    return lower
  else
    greater.left = merge(lower, greater.left)
    return greater
  end
end

def split_binary(orig, value)
  if orig.nil?
    return nil, nil
  elsif orig.x < value
    split_pair = split_binary(orig.right, value)
    orig.right = split_pair[0]
    return orig, split_pair[1]
  else
    split_pair = split_binary(orig.left, value)
    orig.left = split_pair[1]
    return split_pair[0], orig
  end
end

def merge3(lower, equal, greater)
  return merge(merge(lower, equal), greater)
end

class SplitResult
  attr_accessor :lower, :equal, :greater

  def initialize(lower, equal, greater)
    @lower = lower
    @equal = equal
    @greater = greater
  end
end

def split(orig, value)
  lower, equal_greater = split_binary(orig, value)
  equal, greater = split_binary(equal_greater, value + 1)
  return SplitResult.new lower, equal, greater
end

class Tree
  def initialize
    @root = nil
  end

  @root

  def has_value(x)
    splited = split(@root, x)
    res = splited.equal != nil
    @root = merge3(splited.lower, splited.equal, splited.greater)
    return res
  end

  def insert(x)
    splited = split(@root, x)
    splited.equal = Node.new(x) if splited.equal.nil?

    @root = merge3(splited.lower, splited.equal, splited.greater)
  end

  def erase(x)
    splited = split(@root, x)
    @root = merge(splited.lower, splited.greater)
  end
end

def main
  tree = Tree.new
  cur = 5
  res = 0

  (1...1000000).each do |i|
    a = i % 3
    cur = (cur * 57 + 43) % 10007

    case a
    when 0; tree.insert(cur)
    when 1; tree.erase(cur)
    when 2
      res += tree.has_value(cur) ? 1 : 0
    end
  end
  p res
end

startTimeMs = Time.now.to_f * 1000

main()

endTimeMs = Time.now.to_f * 1000
durationMs = (endTimeMs - startTimeMs).to_i

puts "Execution time: #{durationMs}ms"